### DISCUSSION TITLE
Applying [Id] to methods - when? why?

### DISCUSSION THREAD

**Discussion by chwarr** [dotnet/orleans#8835] [Q&A] [answered]
The [`Orleans.IdAttribute`](https://learn.microsoft.com/en-us/dotnet/api/orleans.idattribute) can be applied to a method (it includes `AttributeTargets.Method` in its attribute usage). Why can `[Id]` be applied to a method? What does it mean to serialize a method? Is there some other intended use for `[Id]` on methods?

**Comment by ReubenBond** [comment, marked as answer]
Good question. The answer is that grain methods are serialized, too, when the method is invoked (i.e, RPC). The Id in this case is used an alias to make method call serialization resilient to renames. Alternatively, you can use an `[Alias(x)]` on a method. We merged an analyzer/codefix to automate adding aliases to methods (it's a hint, not a warning).

A serializable class is generated for each method with fields for the method parameters. The invoker's type is identified on the wire using a tuple of the form `("inv", typeof(GrainReference), typeof(IMyInterface), "my-custom-method-id")`
